gobblin-service/src/main/java/org/apache/gobblin/service/modules/flowgraph/Dag.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.flowgraph;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.google.common.collect.Lists;
+
+import org.apache.gobblin.annotation.Alpha;
+
+import lombok.Getter;
+
+
+/**
+ * An implementation of Dag. Assumes that nodes have unique values. Nodes with duplicate values will produce
+ * unpredictable behavior.
+ */
+@Alpha
+@Getter
+public class Dag<T> {
+  private List<DagNode<T>> startNodes;
+  private List<DagNode<T>> endNodes;
```

Comment:
Either start nodes or end nodes are sufficient to get all paths of a dag. We don't need both

Reply:
startNodes and endNodes help:
1. to determine which jobs to start execution with and when a dag execution is complete
2. Useful for concatenating two dags. Just set the parent/child pointers of the endNodes and startNodes.